packages/validator/src/services/doppelgangerService.ts

```diff
@@ -0,0 +1,266 @@
+}
+
+/** Map a DoppelgangerStatus to an integer for rendering in Grafana */
+export const doppelgangerStatusMetrics: {[K in DoppelgangerStatus]: number} = {
+  [DoppelgangerStatus.VerifiedSafe]: 0,
+  [DoppelgangerStatus.Unverified]: 1,
+  [DoppelgangerStatus.Unknown]: 2,
+  [DoppelgangerStatus.DoppelgangerDetected]: 3,
+};
+
+export class DoppelgangerService {
+  private readonly doppelgangerStateByPubkey = new Map<PubkeyHex, DoppelgangerState>();
+
+  constructor(
+    private readonly logger: ILogger,
+    private readonly clock: IClock,
+    private readonly api: Api,
+    private readonly indicesService: IndicesService,
+    private readonly processShutdownCallback: ProcessShutdownCallback,
+    private readonly metrics: Metrics | null
+  ) {
+    this.clock.runEveryEpoch(this.pollLiveness);
+
+    if (metrics) {
+      metrics.doppelganger.statusCount.addCollect(() => this.onScrapeMetrics(metrics));
+    }
+
+    this.logger.info("doppelganger protection enabled", {detectionEpochs: DEFAULT_REMAINING_DETECTION_EPOCHS});
+  }
+
+  registerValidator(pubkeyHex: PubkeyHex): void {
+    const {currentEpoch} = this.clock;
+    // Disable doppelganger protection when the validator was initialized before genesis.
+    // There's no activity before genesis, so doppelganger is pointless.
+    const remainingEpochs = currentEpoch <= 0 ? 0 : DEFAULT_REMAINING_DETECTION_EPOCHS;
+
+    // Log here to alert that validation won't be active until remainingEpochs == 0
+    if (remainingEpochs > 0) {
+      this.logger.info("Registered validator for doppelganger", {remainingEpochs, pubkeyHex});
+    }
+
+    this.doppelgangerStateByPubkey.set(pubkeyHex, {
+      nextEpochToCheck: this.clock.currentEpoch + 1,
+      remainingEpochs,
+    });
+  }
+
+  getStatus(pubKeyHex: PubkeyHex): DoppelgangerStatus {
+    return getStatus(this.doppelgangerStateByPubkey.get(pubKeyHex));
+  }
+
+  isDoppelgangerSafe(pubKeyHex: PubkeyHex): boolean {
+    return getStatus(this.doppelgangerStateByPubkey.get(pubKeyHex)) === DoppelgangerStatus.VerifiedSafe;
+  }
+
+  private pollLiveness = async (currentEpoch: Epoch): Promise<void> => {
+    if (currentEpoch < 0) {
+      return;
+    }
+
+    const endSlotOfCurrentEpoch = computeStartSlotAtEpoch(currentEpoch + 1) - 1;
+    // Run the doppelganger protection check 75% through the last slot of this epoch. This
+    // *should* mean that the BN has seen the blocks and attestations for the epoch
+    await sleep(this.clock.msToSlot(endSlotOfCurrentEpoch + 3 / 4));
+
+    // Collect indices that still need doppelganger checks
+    const pubkeysToCheckWithoutIndex: PubkeyHex[] = [];
+    const indicesToCheck: ValidatorIndex[] = [];
+    const indicesToCheckMap = new Map<ValidatorIndex, PubkeyHex>();
+
+    for (const [pubkeyHex, state] of this.doppelgangerStateByPubkey.entries()) {
+      if (state.remainingEpochs > 0) {
+        const index = this.indicesService.pubkey2index.get(pubkeyHex);
```

Comment:
Suggestion: instead of exposing the internal states of `indicesService` via `pubkey2index` and `pubkey2index` have a method that does the check, they were exposed for